### ISSUE TITLE
In WASM-context: Creating Claims causes Runtime Exception

### ISSUE THREAD

**Issue by niilz** [jedisct1/rust-jwt-simple#6]
Environment:
OS: MacOS BigSur (11.0.1)
Browser: Chrome 86
Project-Setup: built with wasm-pack 0.9.1

jwt-simple version: 0.2.2

I've tried to write a function that creates a JWT.
This function is supposed to be exported via `#[wasm_bidngen]`.
So that it can be used in JavaScript as part of a WebAssembly module.

The function looks something like this:
```
#[wasm_bindgen]
pub fn create_jwt() -> Option<String> {
    // Loading some credentials (here from a const for simplicity)
    let creds: Creds = serde_json::from_str(CREDENTIAL_JSON).unwrap();
    // Creating the encoding key
    let key = RS256KeyPair::from_pem(&creds.private_key);

    // Attempt to create Default-Claims but a Runtime Exception gets thrown
    let claims = claims::Claims::create(Duration::from_hours(2));

    // Do the actual encoding ...
   Some(jwt)
}
````
It compiles fine. And the function is callable on the JavaScript side.
But inside the `Claims::create` function the rust code panics and throws the following RuntimeException:
```
Error importing `index.js`: RuntimeError: unreachable
    at __rust_start_panic (http://localhost:8080/10fe4c26674a9c530e22.module.wasm:wasm-function[13321]:0x336a17)
    at rust_panic (http://localhost:8080/10fe4c26674a9c530e22.module.wasm:wasm-function[12229]:0x32edcd)
    at std::panicking::rust_panic_with_hook::hc3afc981c0e7d6a8 (http://localhost:8080/10fe4c26674a9c530e22.module.wasm:wasm-function[2651]:0x248264)
    at std::panicking::begin_panic::{{closure}}::h75a2b2aafab9823f (http://localhost:8080/10fe4c26674a9c530e22.module.wasm:wasm-function[12209]:0x32eadd)
    at std::sys_common::backtrace::__rust_end_short_backtrace::h80a493e192cc70d9 (http://localhost:8080/10fe4c26674a9c530e22.module.wasm:wasm-function[11817]:0x32a84b)
    at std::panicking::begin_panic::h23e4f4b9c562c80c (http://localhost:8080/10fe4c26674a9c530e22.module.wasm:wasm-function[12208]:0x32eab0)
    at std::sys::wasm::time::SystemTime::now::h820d817d361d7af8 (http://localhost:8080/10fe4c26674a9c530e22.module.wasm:wasm-function[13239]:0x3366e5)
    at std::time::SystemTime::now::h5a8fe1c8bb21e817 (http://localhost:8080/10fe4c26674a9c530e22.module.wasm:wasm-function[13297]:0x33698d)
    at coarsetime::clock::clock_offset::h6a6fea2eece04570 (http://localhost:8080/10fe4c26674a9c530e22.module.wasm:wasm-function[1732]:0x20bc88)
    at core::ops::function::FnOnce::call_once::h49e77ef433e110df (http://localhost:8080/10fe4c26674a9c530e22.module.wasm:wasm-function[11512]:0x326d09)
```

If I understand correctly we cannot use `std::time::Instant::now` in WASM.
But from the error-message, I get the feeling that `std::time::Instant::now` is used and causes the exception.
Is there a way to use **jwt-simple** in a **WebAssembly** module?
Maybe by turning on/off some feature or some other sort of "switch"?

Looling forward to any thoughts on this.

**Comment by jedisct1** [comment]
Without WASI, WebAssembly indeed doesn't have anything to get the current time (why does even Rust silently compile a function of the standard library that can't work?)

As JWT doesn't have any protection against replay attacks, checking token timestamps is quite important. You could get away with it using nonces. But you know what... WebAssembly cannot generate random numbers either. So, no nonce.

You're using RSA. But the Rust implementation of RSA is insecure against side channels, so blinding is required as a workaround. But blinding requires random numbers as well.

In such an environment, I'd recommend using a JavaScript JWT implementation instead. Honestly, that would be faster, smaller and more secure.

If you really have to use Rust, target `wasm32-wasi` instead of `wasm32-unknown-unknown`. I've never used `wasm_bindgen`, but looks like it supports WASI out of the box.

**Comment by jedisct1** [comment]
It actually doesn't compile, so you may not be using that crate.

```text
error: target is not supported, for more information see: https://docs.rs/getrandom/#unsupported-targets
   --> /Users/j/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.2.0/src/lib.rs:224:9
    |
224 | /         compile_error!("target is not supported, for more information see: \
225 | |                         https://docs.rs/getrandom/#unsupported-targets");
    | |_________________________________________________________________________^
```

**Comment by niilz** [comment]
Thank you very much for your quick response and the clarification.
It would indeed be nice if the compilation would already give more insight about what is ok in a wasm context.

Anyhow I guess for client-side JWTs, using a JavaScript library is a helpful suggestion 👍

**Comment by niilz** [comment]
Btw. in the end I was able to create a JWT by using the `base64` and `rsa` crates directly.
And for the `iat` and `exp` values in the `claims` part of the JWT, I used `web_sys::Date::now()`.
This is the raw `wasm-bindgen` import of JavaScript's `Date.now()`.

Since my use case is just a toy-project this is fine for me.
But I would generally be interested whether this approach had similar security implications, like you mentioned before.
